### PR TITLE
enable `socket_vmnet` network driver for qemu Mac M1 ARM64

### PR DESCRIPTION
**Pull request by klaases**
Start minikube with `socket_vmnet` network driver for Mac OS M1 / AARCH64.

- Respect network flag (previously hard-coded to "user").
- Create custom `socket_vmnet_client` and `socket_vmnet` paths.

Overview:
- With default `socket_vmnet_client` and `socket_vmnet` paths:
   - `./out/minikube start --driver=qemu --network=socket`
- With optional custom `socket_vmnet_client` and `socket_vmnet` custom paths:
   - `./out/minikube start --driver=qemu --network=socket --socket-vmnet-client-path=/opt/socket_vmnet/bin/socket_vmnet_client --socket-vmnet-path=/var/run/socket_vmnet`

We are still in the process of implementing finalized socket_vmnet network flags, and will have a follow-up PR shortly.

```
😄  minikube v1.26.1 on Darwin 12.5.1 (arm64)
✨  Using the qemu2 (experimental) driver based on user configuration
❗  Using qemu with --network=socket for 'socket_vmnet' is experimental
👍  Starting control plane node minikube in cluster minikube
🔥  Creating qemu2 VM (CPUs=2, Memory=4000MB, Disk=20000MB) ...

❌  Exiting due to MK_UNIMPLEMENTED: socket_vmnet network flags are not yet implemented with the qemu2 driver.
    See https://github.com/kubernetes/minikube/pull/14890 for details.
```

### CODE REVIEW

File: pkg/drivers/qemu/qemu.go [hunk 1]
```diff
@@ -445,7 +445,9 @@ func (d *Driver) Start() error {
 			d.diskPath())
 	}
 
-	if stdout, stderr, err := cmdOutErr(d.Program, startCmd...); err != nil {
+	socketCmd := append([]string{"/var/run/socket_vmnet", d.Program}, startCmd...)
```
Comment (AkihiroSuda) [inline pkg/drivers/qemu/qemu.go:448]: You have to provide the socket path
https://github.com/lima-vm/socket_vmnet#usage

Comment (klaases) [reply pkg/drivers/qemu/qemu.go:448]: I think we are providing the path?

Client: `"/opt/socket_vmnet/bin/socket_vmnet_client"`
Path: `/var/run/socket_vmnet`

On line 450, we take the client and append the socket path along with other arguments.

Comment (AkihiroSuda) [reply pkg/drivers/qemu/qemu.go:448]: Ah, yes, overlooked “/var/run” sorry

Comment (klaases) [reply pkg/drivers/qemu/qemu.go:448]: Sure, I appreciate your review.

File: polyverse-arm64.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: polyverse
```
Comment (AkihiroSuda) [inline polyverse-arm64.yaml:4]: Seems a irrelevant commit

File: pkg/drivers/qemu/qemu.go [hunk 3]
```diff
@@ -445,7 +445,9 @@ func (d *Driver) Start() error {
...
+	fmt.Printf("socketCmd: %v\n", socketCmd)
+	if stdout, stderr, err := cmdOutErr("/opt/socket_vmnet/bin/socket_vmnet_client", socketCmd...); err != nil {
```
Comment (AkihiroSuda) [inline pkg/drivers/qemu/qemu.go:450]: The paths should be probably configurable.
Also needs documentation.

File: polyverse-arm64.yaml (hunk 2 above)
Comment (klaases) [reply polyverse-arm64.yaml:4]: Yes, this one is just for testing, I was hoping to be able to get a service and/or tunnel running with Minikube + Qemu + socket_vmnet.

I'll remove this before merge.

File: pkg/drivers/qemu/qemu.go (hunk 3 above)
Comment (klaases) [reply pkg/drivers/qemu/qemu.go:450]: That's a good idea, I was hoping to get a minimum viable product working. I'll see if I can make a flag for allowing for a custom `socket_vmnet_client` and `socket_vmnet` path.

I appreciate your review.

File: polyverse-arm64.yaml (hunk 2 above)
Comment (klaases) [reply polyverse-arm64.yaml]: Removed.

File: pkg/drivers/qemu/qemu.go (hunk 3 above)
Comment (klaases) [reply pkg/drivers/qemu/qemu.go]: Updated PR with customizable flag paths for socket_vmnet binaries, please let me know if any more changes are needed.

File: pkg/drivers/qemu/qemu.go [hunk 4]
```diff
@@ -408,6 +409,10 @@ func (d *Driver) Start() error {
 		startCmd = append(startCmd,
 			"-nic", fmt.Sprintf("user,model=virtio,hostfwd=tcp::%d-:22,hostfwd=tcp::%d-:2376,hostname=%s", d.SSHPort, d.EnginePort, d.GetMachineName()),
 		)
+	case "socket":
+		startCmd = append(startCmd,
+			"-nic", fmt.Sprintf("socket,model=virtio,sock=%s", d.NetworkSocket),
```
Comment (klaases) [inline pkg/drivers/qemu/qemu.go:414]: Hi @AkihiroSuda, in your experience, do these parameters for the `-nic` flag look correct for socket_vmnet with minikube socket network?

Comment (AkihiroSuda) [reply pkg/drivers/qemu/qemu.go:414]: Looks good

**Review by spowelljr — CHANGES_REQUESTED** [review]
`d.Network` is hardcoded so it's currently impossible to use `socket`.

When I do manually change `d.Network` to `socket` it fails because of line:
`startCmd = append(startCmd, "-nic", fmt.Sprintf("socket,model=virtio,sock=%s", d.NetworkSocket)`

`d.NetworkSocket` isn't set anywhere in minikube, so it fails with:
`ERROR: qemu-system-aarch64: -nic socket,model=virtio,sock=: Invalid parameter 'sock'`

File: pkg/drivers/qemu/qemu.go (hunk 4 above)
Comment (klaases) [reply pkg/drivers/qemu/qemu.go]: > `d.Network` is hardcoded so it's currently impossible to use `socket`.
> 
> When I do manually change `d.Network` to `socket` it fails because of line: `startCmd = append(startCmd, "-nic", fmt.Sprintf("socket,model=virtio,sock=%s", d.NetworkSocket)`
> 
> `d.NetworkSocket` isn't set anywhere in minikube, so it fails with: `ERROR: qemu-system-aarch64: -nic socket,model=virtio,sock=: Invalid parameter 'sock'`

Appreciate the comment, I have updated the code to reflect the network flag that is being passed into minikube rather than being hardcoded.

For now, if the network flag is empty, will default to the "user" network to maintain current functionality.

Once we have working flags for the qemu socket_vmnet network, will update the default network to "socket" for qemu.

**Comment by klaases** [comment]
> `d.Network` is hardcoded so it's currently impossible to use `socket`.
> 
> When I do manually change `d.Network` to `socket` it fails because of line: `startCmd = append(startCmd, "-nic", fmt.Sprintf("socket,model=virtio,sock=%s", d.NetworkSocket)`
> 
> `d.NetworkSocket` isn't set anywhere in minikube, so it fails with: `ERROR: qemu-system-aarch64: -nic socket,model=virtio,sock=: Invalid parameter 'sock'`

Yes, makes sense. Fixed the hard-coded issue.

For now will default to "user" network to maintain backwards compatibility.

Once we have a set of working qemu + socket_vmnet flags, we'll update the default network to "socket".

File: pkg/drivers/qemu/qemu.go (hunk 4 above)
Comment (klaases) [reply pkg/drivers/qemu/qemu.go]: Hi @AkihiroSuda, we found some errors with these flags. I'll omit them for now but we have a follow-up PR that should more closely express working flags for minikube + qemu + socket_vmnet, appreciate your patience and review.

**Review by spowelljr — APPROVED** [review]
